tests/kernel/timer/timer_api/src/main.c

```diff
@@ -226,18 +227,20 @@ void test_timer_periodicity(void)
 	 * whatever duration that was set.
 	 */
 	k_timer_status_sync(&timer);
-	tdata.timestamp = k_uptime_get();
+	cyc0 = k_cycle_get_32();
 
 	for (int i = 0; i < EXPIRE_TIMES; i++) {
 		/** TESTPOINT: expired times returned by status sync */
 		TIMER_ASSERT(k_timer_status_sync(&timer) == 1, &timer);
 
-		delta = k_uptime_delta(&tdata.timestamp);
+		now = k_cycle_get_32();
+		int dticks = (now - cyc0) / sys_clock_hw_cycles_per_tick();
 
-		/** TESTPOINT: check if timer fired within 1ms of the
-		 *  expected period (firing time)
+		/** TESTPOINT: check if timer fired within two ticks
+		 *  of the expected period (firing time)
 		 */
-		TIMER_ASSERT(WITHIN_ERROR(delta, PERIOD, 1), &timer);
+		TIMER_ASSERT(abs(dticks - _ms_to_ticks(PERIOD)) <= 2, &timer);
```

Comment:
Is firing before the timeout correct? IMHO We can fire a bit later, but we should detect the situation when timer fired too early.

Reply:
It's not that it fires before the timeout, it's that the internal computation of when the timeout is going to expire inherently races against the test code's, so if you compute the timestamp before the k_timer_start() call, you don't know that they aren't on opposite sides of the same tick (and vice versa).  This used to be mostly academic as timeout handling always happened out of callbacks (or due to wakeups from callbacks) that only happened on tick boundaries, and the z_tick_get() value would change only on those same boundaries.  But in tickless mode it's realtime.

That said: I'm finding lots and lots of places in the tests that are bitten by this kind of aliasing (you can see the patches in the timerhacks branch), and intuitively they seem to be more common than they should.  For simple timer test code, it SHOULD be possible to have the tickless ordering be identical to ticked, and I'm worried that the fact I have to change so many tests is really evidence of a bug in the systick driver in tickless mode.

I'm removing this particular patch from this series, basically.